providers/class-two-factor-webauthn.php

```diff
@@ -0,0 +1,642 @@
+<?php
+/**
+ * Class for creating WebAuthn token
+ *
+ * @package Two_Factor
+ */
+
+/**
+ * Class Two_Factor_WebAuthn.
+ */
+class Two_Factor_WebAuthn extends Two_Factor_Provider {
+
+	/**
+	 * The user meta login key.
+	 *
+	 * @type string
+	 */
+	const LOGIN_USERMETA = '_two_factor_webauthn_login';
+
+	/**
+	 * Handling WebAuthn requests and responses.
+	 *
+	 * @var WebAuthnHandler
+	 */
+	protected $webauthn;
+
+	/**
+	 * Holding a users keys.
+	 *
+	 * @var WebAuthnKeyStore
+	 */
+	protected $key_store;
+
+	/**
+	 * Ensures only one instance of this class exists in memory at any one time.
+	 *
+	 * @return \Two_Factor_WebAuthn
+	 */
+	public static function get_instance() {
+		static $instance;
+
+		if ( ! isset( $instance ) ) {
+			$instance = new self();
+		}
+
+		return $instance;
+	}
+
+	/**
+	 * Class constructor.
+	 */
+	protected function __construct() {
+
+		require_once TWO_FACTOR_DIR . 'includes/WebAuthn/class-webauthn-handler.php';
+		require_once TWO_FACTOR_DIR . 'includes/WebAuthn/class-cbor-decoder.php';
+		require_once TWO_FACTOR_DIR . 'includes/WebAuthn/class-webauthn-keystore.php';
+
+		$this->webauthn = new WebAuthnHandler( $this->get_app_id() );
+
+		$this->key_store = WebAuthnKeyStore::instance();
+
+		wp_register_script(
+			'webauthn-login',
+			plugins_url( 'providers/js/webauthn-login.js', dirname( __FILE__ ) ),
+			array( 'jquery' ),
+			TWO_FACTOR_VERSION,
+			true
+		);
+
+		wp_register_script(
+			'webauthn-admin',
+			plugins_url( 'providers/js/webauthn-admin.js', dirname( __FILE__ ) ),
+			array( 'jquery' ),
+			TWO_FACTOR_VERSION,
+			true
+		);
+
+		wp_register_style(
+			'webauthn-admin',
+			plugins_url( 'providers/css/webauthn-admin.css', dirname( __FILE__ ) ),
+			array(),
+			TWO_FACTOR_VERSION
+		);
+
+		wp_register_style(
+			'webauthn-login',
+			plugins_url( 'providers/css/webauthn-login.css', dirname( __FILE__ ) ),
+			array(),
+			TWO_FACTOR_VERSION
+		);
+
+		add_action( 'wp_ajax_webauthn-register', array( $this, 'ajax_register' ) );
+		add_action( 'wp_ajax_webauthn-edit-key', array( $this, 'ajax_edit_key' ) );
+		add_action( 'wp_ajax_webauthn-delete-key', array( $this, 'ajax_delete_key' ) );
+		add_action( 'wp_ajax_webauthn-test-key', array( $this, 'ajax_test_key' ) );
+
+		add_action( 'two_factor_user_options_' . __CLASS__, array( $this, 'user_options' ) );
+
+		parent::__construct();
+
+	}
+
+
+	/**
+	 * Enqueue assets for login form.
+	 */
+	public function login_enqueue_assets() {
+		wp_enqueue_script( 'webauthn-login' );
+		wp_enqueue_style( 'webauthn-login' );
+	}
+
+	/**
+	 * Return the U2F AppId. WebAuthn requires the AppID
+	 * to be the current domain or a suffix of it.
+	 *
+	 * @return string AppID FQDN
+	 */
+	public function get_app_id() {
+
+		$fqdn = wp_parse_url( network_site_url(), PHP_URL_HOST );
```

Comment:
It is probably not critical, but WebAuthn requires an origin here, not just the domain. That is if the webserver for some reason uses a non-standard port for HTTPS (say, 8443), it needs to be included in the AppID.

The best way is probably to copy the logic from [`get_u2f_app_id()`](https://github.com/WordPress/two-factor/blob/e72192eef199285e8edc954e049ee888789179d8/providers/class-two-factor-fido-u2f.php#L103) (just omit the protocol).